### PR TITLE
[move] Add CFGIR Ast Visitor. Clean up rough edges of T::Program.

### PR DESCRIPTION
**Pull request by tnowacki**
## Description 

- Some lints probably want to use the CFGIR, since they are easier to do post optimizations 
- Cleaned up the linter APIs

## Test plan 

- internal change 

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: external-crates/move/crates/move-compiler/src/cfgir/translate.rs [hunk 1]
```diff
@@ -983,17 +986,22 @@ fn destructure_tuple<T, U>((fst, snd): &(T, U)) -> (&T, &U) {
...
+    }
+
+    for visitor in &context.env.visitors().cfgir {
+        let mut v = visitor.borrow_mut();
+        v.visit(context.env, prog)
     }
 }
 
-fn visit_module(
+fn absint_visit_module(
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/cfgir/translate.rs:1004]: Could we rewrite this to use the new visitor?

File: external-crates/move/crates/move-compiler/src/cfgir/visitor.rs [hunk 2]
```diff
@@ -41,7 +55,247 @@ pub trait AbstractInterpreterVisitor {
...
+            | Command_::IgnoreAndPop { exp: e, .. }
+            | Command_::JumpIf { cond: e, .. }
+            | Command_::VariantSwitch { subject: e, .. } => {
+                self.visit_exp(e);
+            }
+            Command_::Mutate(el, er) => {
+                self.visit_exp(el);
+                self.visit_exp(er);
+            }
+            Command_::Break(_) | Command_::Continue(_) | Command_::Jump { .. } => (),
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/cfgir/visitor.rs:242]: I wonder if these could report an error somehow.

File: external-crates/move/crates/move-compiler/src/cfgir/visitor.rs [hunk 3]
```diff
@@ -41,7 +55,247 @@ pub trait AbstractInterpreterVisitor {
...
+        &mut self,
+        module: ModuleIdent,
+        constant_name: ConstantName,
+        cdef: &mut G::Constant,
+    ) {
+        self.add_warning_filter_scope(cdef.warning_filter.clone());
+        if self.visit_constant_custom(module, constant_name, cdef) {
+            self.pop_warning_filter_scope();
+            return;
+        }
```
Comment (cgswords) [inline external-crates/move/crates/move-compiler/src/cfgir/visitor.rs:176]: Do we want to add `visit_value` and call that here?

**Review by cgswords — APPROVED** [review]
LGTM modulo a few small questions.

File: external-crates/move/crates/move-compiler/src/cfgir/translate.rs (hunk 1 above)
Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/cfgir/translate.rs:1004]: Oh thats fun. Stuffing all of the absint visitors in a visitor?

File: external-crates/move/crates/move-compiler/src/cfgir/visitor.rs (hunk 2 above)
Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/cfgir/visitor.rs:242]: Sure

File: external-crates/move/crates/move-compiler/src/cfgir/visitor.rs (hunk 3 above)
Comment (tnowacki) [reply external-crates/move/crates/move-compiler/src/cfgir/visitor.rs:176]: Sure, seems helpful in the CFGIR

**Pull request merged**